### ISSUE TITLE
AcitivityPlanner: JSON serialization issue

### ISSUE THREAD

**Issue by nschuessler** [microsoft/semantic-kernel#1919]
**Describe the bug**
ActivityPlanner `CreatePlanAsync` depends on `SKContext.ToString()` to create valid JSON for parsing.  `.ToString()` is also used for debugging and user interaction so this is a fragile implementation. In fact, it currently fails:

**To Reproduce**
1. Unpack the attached .zip file with the repro solution. It has two skills and a main program.
2. Update the key and instance to point to a davinci-003
3. Build and run
4. You will get a `PlanException` with inner exception JsonSerializationException

This is because the `SKContext.ToString()` generates:

```
Variables = 1, Input = together.
 
 {
    "plan":{
        "rationale": "the list contains a function that allows to add two numbers together",
        "function": "_GLOBAL_FUNCTIONS_.Add",
        "parameters": {
            "number1": 4,
             "number2": 5
        }
   }
}
, Skills = 3, Culture = English(United States)
```

**Expected behavior**
The plan is created correctly and produces the sum of 4 and 5.
A better design might be to have a ToJson() method on SKContext that is guaranteed to generate valid JSON.

**Screenshots**

**Desktop (please complete the following information):**
 - OS: Windows
 - IDE: Visual Studio 17.6.x
 - NuGet Package Version [e.g. 0.1.0]

**Additional context**
Zip file with solution
[ActivityPlannerRepro.zip](https://github.com/microsoft/semantic-kernel/files/12002628/ActivityPlannerRepro.zip)

**Comment by nschuessler** [comment]
Looks like the code has been changed, but hasn't been released it has the following now:

     `  ActionPlanResponse? planData = this.ParsePlannerResult(result);`

**Comment by nschuessler** [comment]
nit: In `ParsePlannerResult` the `Regex` expression is a constant string so should be a static property (allocated once) and have the `Compiled` regex option specified for performance.  Creating and running plans will happen frequently for server applications so we should be optimal about it:

```
 private ActionPlanResponse? ParsePlannerResult(SKContext plannerResult)
 {
     Regex planRegex = new("^[^{}]*(((?'Open'{)[^{}]*)+((?'Close-Open'})[^{}]*)+)*(?(Open)(?!))", RegexOptions.Singleline);
     Match match = planRegex.Match(plannerResult.ToString());
```

**Comment by nacharya1** [comment]
Closing since we merged #1954
